### DISCUSSION TITLE
How to receive / process DynamoDB Triggers?

### DISCUSSION THREAD

**Discussion by joeuelk** [soto-project/soto#620] [Q&A]
Does anyone know how to receive a DynamoDB trigger?

The example below works fine with the APIGateway but I can't get it to process DDB Triggers.

I've verified data is being sent to this Lambda when I write to a table that has a stream setup but I am getting an error saying it can't decode the data because the `resource` coding key is missing.

I've tried changing the `event` and the `In` values to various things but nothing is working.

Thanks

```
struct LamdaHandler: EventLoopLambdaHandler {
    
    typealias In = APIGateway.Request
    typealias Out = APIGateway.Response
    
    func handle(context: Lambda.Context, event: APIGateway.Request) -> EventLoopFuture<APIGateway.Response> {
        guard let handler = Handler.current else {
            return context.eventLoop.makeSucceededFuture(APIGateway.Response(with: APIError.requestError, statusCode: .badRequest))
        }
    }

}
```

**Comment by adam-fowler** [comment]
This question should probably be asked in the `swift-aws-lambda-runtime` repo, but I can see if I can help. First it looks like you are using an old version of the lambda runtime. It hasn't had a release in a while, but has had quite a few updates like moving all the event types to `swift-aws-lambda-events` and async/await support. 

DynamoDB triggered lambdas should use `DynamoDBEvent` from the `swift-aws-lambda-events` repo as their `In` type. eg
```swift
struct LamdaHandler: EventLoopLambdaHandler {
    typealias In = DynamoDBEvent
    typealias Out = Void
    
    func handle(context: Lambda.Context, event: APIGateway.Request) -> EventLoopFuture<APIGateway.Response> {
        guard let handler = Handler.current else {
            return context.eventLoop.makeSucceededVoidFuture())
        }
    }
}
```
If that doesn't work then there is most likely an error in the `DynamoDBEvent` type and an issue should be raised in the events repo.

**Comment by joeuelk** [comment]
I think I'm getting confused about the difference between `soto` and the `swift-aws-lambda-runtime`. Is there overlap? Does one replace the other? Is there an equivalent `EventLoopLambdaHandler` within Soto?

Thank you!

**Reply by 0xTim** [reply]
Soto is an SDK for calling AWS services via their API. The AWS Lambda runtime allows you to run Swift code in lambdas. There's no overlap between the two
